src/core/common/timer.cpp

```diff
@@ -181,39 +173,16 @@ Ip6::Ip6 *TimerScheduler::GetIp6(void)
     return Ip6::Ip6FromTimerScheduler(this);
 }
 
-bool TimerScheduler::TimerCompare(const Timer &aTimerA, const Timer &aTimerB)
+bool TimerScheduler::IsStrictlyBefore(uint32_t aTimeA, uint32_t aTimeB)
 {
-    uint32_t now = otPlatAlarmGetNow();
-    uint32_t elapsedA = now - aTimerA.mT0;
-    uint32_t elapsedB = now - aTimerB.mT0;
-    bool retval = false;
-
-    if (aTimerA.mDt >= elapsedA && aTimerB.mDt >= elapsedB)
-    {
-        uint32_t remainingA = aTimerA.mDt - elapsedA;
-        uint32_t remainingB = aTimerB.mDt - elapsedB;
-
-        if (remainingA < remainingB)
-        {
-            retval = true;
-        }
-    }
-    else if (aTimerA.mDt < elapsedA && aTimerB.mDt >= elapsedB)
-    {
-        retval = true;
-    }
-    else if (aTimerA.mDt < elapsedA && aTimerB.mDt < elapsedB)
-    {
-        uint32_t expiredByA = elapsedA - aTimerA.mDt;
-        uint32_t expiredByB = elapsedB - aTimerB.mDt;
+    uint32_t diff = aTimeA - aTimeB;
 
-        if (expiredByB < expiredByA)
-        {
-            retval = true;
-        }
-    }
+    // Three cases:
+    // 1) aTimeA is before  aTimeB  =>  Difference is negative (last bit of difference is set)   => Returning true.
+    // 2) aTimeA is same as aTimeB  =>  Difference is zero     (last bit of difference is clear) => Returning false.
+    // 3) aTimeA is after   aTimeB  =>  Difference is positive (last bit of difference is clear) => Returning false.
 
-    return retval;
+    return ((diff & (1UL << 31)) != 0);
```

Comment:
I was thinking of this actually. Did a quick research on how `static_cast<int32_t>` would convert from `uint32_t` to `int32_t`... Seems like this may be technically undefined behavior and can be platform/compiler dependent. I felt checking the last bit directly would be safer and always correct.